tests/phpunit/CRM/Civicase/Hook/ValidateForm/SendBulkEmailTest.php

```diff
@@ -0,0 +1,190 @@
+<?php
+
+use CRM_Civicase_Hook_ValidateForm_SendBulkEmail as SendBulkEmailHook;
+
+/**
+ * Test class for the CRM_Civicase_Hook_ValidateForm_SendBulkEmail.
+ *
+ * @group headless
+ */
+class CRM_Civicase_Hook_SendBulkEmailTest extends BaseHeadlessTest {
+
+  use CRM_Civicase_Helpers_SessionTrait;
+
+  /**
+   * Instance of the form sent on the hook.
+   *
+   * @var CRM_Core_Form
+   */
+  private $form;
+
+  /**
+   * Case Type created for using on tests.
+   *
+   * @var array
+   */
+  private $caseType;
+
+  /**
+   * {@inheritDoc}
+   */
+  public function setUp() {
+    parent::setUp();
+
+    $loggedInUser = CRM_Civicase_Test_Fabricator_Contact::fabricateWithEmail();
+    $this->registerCurrentLoggedInContactInSession($loggedInUser['id']);
+
+    $this->caseType = CRM_Civicase_Test_Fabricator_CaseType::fabricate()['id'];
+  }
+
+  /**
+   * Test that the hook is executed when we use an email form.
+   */
+  public function testHookRunsWhenEmailFormIsUsed() {
+    $contacts = [];
+    $contacts[] = CRM_Civicase_Test_Fabricator_Contact::fabricateWithEmail()['id'];
+    $cases = [];
+    $cases[] = $this->createCaseForContact($contacts[0])['id'];
+
+    $_REQUEST['caseRoles'] = $_GET['caseRoles'] = 'client';
+
+    $this->form = new CRM_Contact_Form_Task_Email();
+    $result = $this->runHook($cases, $contacts);
+
+    $this->assertTrue($result);
+  }
+
+  /**
+   * Test that the hook is not executed when we use another form.
+   */
+  public function testHookDoesNotRunWhenOtherFormIsUsed() {
+    $contacts = [];
+    $contacts[] = CRM_Civicase_Test_Fabricator_Contact::fabricateWithEmail()['id'];
+    $cases = [];
+    $cases[] = $this->createCaseForContact($contacts[0])['id'];
+
+    $_REQUEST['caseRoles'] = $_GET['caseRoles'] = 'client';
+
+    $this->form = new CRM_Contact_Form_Task();
+    $result = $this->runHook($cases, $contacts);
+
+    $this->assertFalse($result);
+  }
+
+  /**
+   * Test that the hook is not executed when no cases are selected.
+   */
+  public function testHookDoesNotRunWhenNoCasesAreSelected() {
+    $contacts = [];
+    $cases = [];
+
+    $_REQUEST['caseRoles'] = $_GET['caseRoles'] = 'client';
+
+    $this->form = new CRM_Contact_Form_Task_Email();
+    $result = $this->runHook($cases, $contacts);
+
+    $this->assertFalse($result);
+  }
+
+  /**
+   * Verify that one email activity is created for every case involved.
+   */
+  public function testHookCreateEmailActivitiesForCases() {
```

Comment:
I added 3 more tests: https://github.com/compucorp/uk.co.compucorp.civicase/pull/804/commits/536ab50163540394dce6c42b28be6662c9f5a82d
- for checking that excluded contacts are not notified (related to your previous comment about the interface change)
- for sending only to clients
- for sending only to case roles
Honestly, I am not happy with the test that I did. The main purpose of this class is to modify the number of times that form->submit is called, then I should check that, directly or indirectly.
The more direct way is to check the number of emails sent, but that was complicated and it would require investigation.
The email activity stores somewhere the information about all the contacts involved, but I couldn't find a way to be sure which ones were the recipients. I checked database and code for that.
I think that checking the existence or not of the activity provides some certainty, enough for this task (and budget), but can be improved later. Let me know if you think that is ok.

Reply:
We can actually check that emails were sent. 
First we need to redirect emails to the database: https://github.com/compucorp/civihr/pull/1743/files#diff-c7e0269a72907073b238b18a2c95eb2eb2f4f06d134404de26103edfe6a4aeceR8
Then add methods for checking emails when a mail is sent and deleting emails after the test has ran: https://github.com/compucorp/civihr/blob/03da625c3bea5c9d6ca35b966787cda6d7eba31b/uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/helpers/MailHelpersTrait.php#L35-L47
You can find full details on this PR: https://github.com/compucorp/civihr/pull/1743

Reply:
@tunbola I followed you suggestion, and the test class is almost rewritten using that approach: https://github.com/compucorp/uk.co.compucorp.civicase/pull/804/commits/06d08bdd232e0bd60250069e7fef6d3c4934e090
I also pushed this other change, for a fix that I discovered doing these tests: https://github.com/compucorp/uk.co.compucorp.civicase/pull/804/commits/57f27eb4b30d9391325da8872ab3c02b29ece1f8